File: apps/core/lib/core/task.ex
defmodule Core.Task do
  defstruct [id: nil, title: "", done: false]

  def build(title) do
    %Core.Task{id: Core.ID.generate(), title: title}
  end

  def toggle(%{done: true} = task),   do: uncomplete(task)
  def toggle(%{done: false} = task),  do: complete(task)

  def complete(task),   do: Map.put(task, :done, true)
  def uncomplete(task), do: Map.put(task, :done, false)
end

File: apps/core/lib/core/task.rb
class Task
  attr_reader :id, :title, :done

  def initialize(title)
    @id     = ID.generate
    @title  = title
    @done   = false
  end

  def toggle
    done ? uncomplete : complete
  end

  def complete
    @done = true
  end

  def uncomplete
    @done = false
  end
end
